packages/gatsby-source-contentful/src/on-pre-bootstrap.js

```diff
@@ -0,0 +1,261 @@
+// @todo import syntax!
+const isOnline = require(`is-online`)
+const _ = require(`lodash`)
+const fs = require(`fs-extra`)
+const path = require(`path`)
+const v8 = require(`v8`)
+const { CODES } = require(`./report`)
+const fetchData = require(`./fetch`)
+const { createPluginConfig } = require(`./plugin-options`)
+
+export async function onPreBootstrap(
```

Comment:
@wardpeet we would want to make sure with v4 that this function is `onPluginInit`. Otherwise the workers with source nodes and createSchemaCustomization will have the wrong data, correct?

Reply:
It does a lot of tricky things that I don't understand why we're doing it in onPreBootstrap or would even do it in onPluginInit.

Why are we fetching data here?

Reply:
Yes, if you don't call this code, you will stay with the old Contentful data

Reply:
I understand but why can't we put this all in sourcenodes?

Reply:
@wardpeet I need the content types to be available in `createSchemaCustomization`

The next "major" #31385 will require even more data from Contentful to available when creating the schema.

Reply:
I think we need to do something like this https://github.com/gatsbyjs/gatsby/commit/73ae944de8d5cea2198731921eb0404ad16460c8# where we create a function to fetch data and to fetch types, so we can fetch types solely in `createShemaCustomization` and fetch data solely in `sourceNodes`. Right now you're fetching both data and types, and with v4 all the processes are seperate, so if we keep it this way we'd have to fetch both data and types in both `createSchemaCustomization` and `sourceNodes`

Reply:
It might be possible to split up the huge fetch function into two functions:

`createSchemaCustomization` will load content types
`sourceNodes` will load content types + all of the rest

Loading data twice is bad, thats why I originally moved all of this into bootstrap. Could work around this with the Gatsby cache, but only if I can rely on the call order of `createSchemaCustomization` && `sourceNodes`

Reply:
Why can't we call contentTypes separate from data? Doesn't contentful supports that? Relying on cache will be problematic when the process crashes - you don't really know where we left off.

Reply:
We can, it is supported. I just wonder If i can avoid calling contentTypes twice.

Will try it now, let u know where I end up :)

Reply:
`GATSBY_CONTENTFUL_EXPERIMENTAL_FORCE_CACHE` is going to make it tricky, as it assumes we have all the data in one place which will not be true anymore.

I might store the content type information in a separate file 🤔

Reply:
I got a variant working in e2e and manual tests. Unit tests are green except the huge snapshot based tests. Will fix these after lunch, then we should be able to do another review round :)